### PR TITLE
[#266] Edge-volume model

### LINKED ISSUE #266: Edge-based volume for place_on_edges()

### PR DESCRIPTION
**Pull request by susanw1**
Added src/polysymmetrica/core/edge_regions.scad, with `ps_edge_region_shells(...)` and ps_edge_region_volume(...). The new approach derives edge-region atom shells from the two adjacent faces’ filled boundary spans, so anti-truncated tetrahedron cases can split an edge into multiple volumes instead of forcing one cuboid corridor.

  Also added focused tests in `src/tests/core/TestEdgeRegions.scad`, registered them in `src/tests/run_all.scad`, documented the new edge-region helpers in `docs/reference/placement_data_model.md`, and added a visual/example playground at `src/examples/basics/main_edge_region_volume.scad`.

### CODE REVIEW

File: src/polysymmetrica/core/edge_regions.scad [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+function _ps_er_atom_shell(edge_site, atom, span0, span1, outset, z0, z1, eps=1e-8) =
+    let(
+        edge_len = ps_edge_site_edge_len(edge_site),
+        x0 = -edge_len / 2 + atom[0] * edge_len,
+        x1 = -edge_len / 2 + atom[1] * edge_len,
+        ray0 = _ps_er_span_filled_ray_edge_local(span0, edge_site, eps),
+        ray1 = _ps_er_span_filled_ray_edge_local(span1, edge_site, eps),
+        bottom_ys = _ps_er_oriented_side_ys(ray0, ray1, outset, z0, eps),
+        top_ys = _ps_er_oriented_side_ys(ray0, ray1, outset, z1, eps),
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/edge_regions.scad]: Fixed I think.

@codex review

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/edge_regions.scad]: • I’ll fix this in the edge-region shell construction rather than trying to paper over it at the call site. The right
  invariant is that each side wall must stay attached to the same adjacent-face constraint from bottom to top; if those
  constraints swap order across Z, that atom needs to split at the crossing Z.
• I’m going to replace the per-plane sorting helper with “ordered constraints”: derive each side as a named linear
  constraint, decide left/right once across the whole Z span, and assert if the two named constraints cross or collapse
  within that span. That keeps the generated side quads attached to their source face.
• I’m adding a focused positive test for stable identity ordering and a negative test for the crossing case. The
  negative one is useful here because the desired behavior is an explicit assertion, not a silent geometry output.

Fixed the review issue in src/polysymmetrica/core/edge_regions.scad: side constraints are now kept as named linear constraints across the whole Z span, ordered once, and rejected if they cross/collapse between z0 and z1. That prevents bottom/top loops from sorting independently and connecting the wrong adjacent-face sides.

  Added coverage in src/tests/core/TestEdgeRegions.scad plus a negative assertion test at src/tests/negative/ edge_region_side_constraints_cross.scad.

• I’m checking the loop-shell renderer before deciding the exact split representation. If it only supports same-arity
  cap loops, I’ll split around the crossing by a tiny tolerance; if it can accept a collapsed cap cleanly, I’ll keep the
  two sub-volumes touching exactly.
• The least awkward representation here is a real split: ordinary frusta when both caps have width, and a closed wedge when a subrange collapses to the crossing line. That avoids a fake gap at the crossing and avoids connecting sides to the wrong face.
• I spotted a bad refactor detail before running it: the stable subrange path was about to recompute side constraints
  with a dummy outset. I’m correcting that so split shells use the exact constraints already derived from the adjacent
  faces.
• The public ps_edge_region_shells(...) path now returns all sub-shells for an edge atom. I added a regression matching the failing example parameters against the anti-tetrahedron so this exact class of crossing no longer sneaks through as an example-only failure.
• One cleanup before I finish: the earlier single-shell private helper is now stale because public code always goes
  through the splitting path. I’m removing that unused branch and the negative test that only exercised the stale
  helper.
• Fixed. The example failure was real: crossing side constraints need splitting, not assertion.
  I changed src/polysymmetrica/core/edge_regions.scad so an edge atom can split across z at the side-constraint crossing. Stable subranges emit normal loop shells; subranges that collapse at the crossing emit a closed wedge shell, so side identity is preserved without a fake connection.

File: src/polysymmetrica/core/edge_regions.scad [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+                [3, 2, 5, 4],
+                [0, 3, 4],
+                [1, 5, 2]
+            ]
+            : [
+                [2, 3, 4, 5],
+                [0, 1, 3, 2],
+                [0, 5, 4, 1],
+                [0, 2, 5],
+                [1, 4, 3]
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/edge_regions.scad]: Changed the upper-wide wedge branch in `src/polysymmetrica/core/edge_regions.scad` by reversing the face lists, matching LHR-outward orientation. Added a signed-volume regression in `src/tests/core/TestEdgeRegions.scad` to confirm crossing wedge shells are LHR-outward.

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/edge_regions.scad]: @codex review

File: src/polysymmetrica/core/edge_regions.scad [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+//   edge_idx = source edge index; defaults to `$ps_edge_idx` inside `place_on_edges(...)`.
+//   mode = face fill mode used when deriving adjacent face boundary spans.
+//   eps = geometric tolerance.
+function ps_edge_region_shells(poly, outset, z0, z1, inter_radius=1, edge_len=undef, edge_idx=$ps_edge_idx, mode="nonzero", eps=1e-8) =
+    let(
+        _edge_idx = assert(!is_undef(edge_idx), "ps_edge_region_shells: edge_idx is required; call inside place_on_edges(...) or pass edge_idx"),
+        _outset = assert(outset > eps, "ps_edge_region_shells: outset must be positive"),
+        _z0 = assert(!is_undef(z0), "ps_edge_region_shells: z0 must be defined"),
+        _z1 = assert(!is_undef(z1), "ps_edge_region_shells: z1 must be defined"),
+        _height = assert(abs(z1 - z0) > eps, "ps_edge_region_shells: z0 and z1 must differ"),
```
[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/edge_regions.scad]: Changed `src/polysymmetrica/core/edge_regions.scad:302` so ps_edge_region_shells() now asserts `z1 > z0 + eps`, rather than accepting reversed bounds.

  Added negative coverage in `src/tests/negative/edge_region_reversed_z_bounds.scad:1`.

[the remark this answers is not in the document]
Comment (susanw1) [reply src/polysymmetrica/core/edge_regions.scad]: @codex review

**Comment by susanw1** [comment]
#### Claude says:

Good — that's the real, clean diff for this PR: a new `edge_regions.scad` module (350 lines) plus tests, docs, and an example. Let's read it in full.Let me trace how this evolved commit-by-commit, since the wedge/z-crossing logic is the most intricate part and I want to see what broke along the way.I pulled the PR branch and read `edge_regions.scad` in full, plus walked the commit history to see what broke along the way — that's usually the most informative part of a review like this.

## The architecture is right, and it's the thing we sketched

`_ps_er_face_edge_boundary_spans` pulls each adjacent face's *own* already-computed boundary-span sites (literally reusing `_ps_face_boundary_span_sites`, the same call `face_regions.scad` makes internally) and filters to the ones touching this edge, canonicalizing each into a global `[t0,t1]` range along the edge regardless of which direction that face's own loop runs. `_ps_er_atom_intervals` unions the transition points from *both* faces' span lists and cuts the edge into intervals at every point where either side's atom membership changes. That's exactly "derive the edge's atoms from the two faces' own decomposition, don't reinvent it" — and the PR body says as much directly. Good, this is the real fix, not a bigger, more complicated version of the same underpowered idea.

Nicer than I'd have designed it myself: the per-atom "which way is outward" direction (`_ps_er_span_filled_ray_edge_local`) is pulled straight from each contributing face's already-correct `filled_side`, transformed face-local → parent → edge-local. That means the edge system never needs its own winding/exposure classification pass the way `face_regions.scad` does (`_ps_fr_cell_winding_signs` and friends) — it inherits correctness from the faces instead of re-deriving it. Consistent with why every emitted shell here leaves `exposure_sign` as `undef`: there's no separate top/bottom question to answer at the edge level, the orientation was already settled per-span before it got here. Worth being a deliberate, understood choice rather than an oversight if anyone goes looking for it later — a future caller wanting to distinguish "ordinary rod material" from some atet-style exceptional case can't currently ask the shell object for that, only infer it from which spans fed the atom.

## The bug that got fixed is a real one, and a subtle one

Diffing the first commit against the "preserve side identity" fix: the original `_ps_er_oriented_side_ys` computed each side's Y position independently at z0 and z1, then just sorted `y0<=y1 ? [y0,y1] : [y1,y0]` at *each* Z level separately. That's fine when the two side constraints stay parallel, but for a converging cross-section (exactly what an acute atet dihedral produces), the two rays can cross somewhere between z0 and z1 — meaning "the smaller Y value" at z0 might belong to face A, but at z1 it belongs to face B. Sorting independently at each end loses track of which physical face each wall actually came from, and lofting a bottom quad to a top quad by position rather than by identity would produce a twisted/self-intersecting extrusion right where the two faces swap sides. The fix computes the crossing Z once, splits the atom into stable sub-ranges at that point, and fixes the left/right assignment once per sub-range from its midpoint rather than re-deriving it at both ends. That's a real degenerate-geometry bug, not a nitpick, and it's the kind of thing that's very easy to get subtly wrong on a first pass — glad it got caught before merge rather than after.

The wedge shell for the degenerate end (one side's width collapsing to ~0 exactly at the crossing point) is the right call structurally — a proper 6-point tent/wedge primitive with two triangular end caps and three quad sides, rather than trying to force a zero-width quad through the general loop-shell path. I checked the connectivity by hand (the two triangle caps `[0,3,4]`/`[1,5,2]` correctly use the two "wide" corners plus the "tip" point that share the same edge-length position, and the three side quads correctly stitch wide-to-tip along each of the two lateral faces) — that part's sound. The actual winding direction of the mirrored ("wide-on-top") branch is not something I'd trust myself to verify by eye, and neither did they: the fix (`ffe82d1`) came with `test_ps_edge_region_shells__crossing_wedges_are_lhr_outward`, which sweeps every edge of a `t=-1` atet with parameters chosen specifically to force wedge cases, and checks every resulting 6-point shell against `_ps_faces_signed_volume6_rhr` — the same signed-volume primitive from the orientation-robustness work several PRs back. That's exactly the right way to close out this class of bug, and reusing the established ground-truth primitive instead of inventing a new ad hoc check is good discipline. I'd trust that test over my own manual trace of the winding here.

`5c5c5dc` tightening `abs(z1-z0)>eps` to `z1>z0+eps` with a proper negative test is minor but continues the same "fail loudly with a clear message" house style this whole project has kept up since the very first vertex-fan work.

## Test coverage

Good breadth: a plain single-atom tetrahedron edge as the baseline, a `place_on_edges` context-integration test (confirms `$ps_edge_idx`/`$ps_edge_len` thread through correctly), and — nicely — the atet split-count test reuses the *exact same* `t=-0.5` fixture the face-region tests used for #209, so the analogy between "faces split into multiple shells" and "edges split into multiple atoms" is validated on literally the same stress case rather than a new one invented for edges. Plus clean isolated unit tests of `_ps_er_side_ys_for_z` and `_ps_er_stable_z_ranges` on synthetic constraint values, which is the right way to pin down the crossing-split logic without needing a real polyhedron in the loop.

The example file is a good sanity gallery too — atet front and center, plus a plain Platonic, an Archimedean, a Catalan (non-regular faces), and an irregularly-parametrized truncated dodecahedron, so it's not just validated against the one adversarial fixture.

## What I'd want before or shortly after merge

- The doc's stated limitation — "closed two-face edges only; endpoint handoff to vertex-node volumes is not applied here" — is honest and exactly the boundary we talked through for #264: this deliberately doesn't try to solve the vertex-node problem, it just stops short and leaves a rod that runs to the raw edge endpoints. That's the right scope for this PR. Worth making sure the follow-up vertex-node issue explicitly says it needs to come back and clip these rods rather than assuming they're already final.
- Given how non-obvious the wedge winding bug was — even the implementation needed an automated volume check rather than getting it right by inspection — I'd add a short comment right at `_ps_er_atom_wedge_shell` flagging "winding here is validated by the signed-volume sweep in TestEdgeRegions; re-run it if you touch this branch." Cheap insurance against a future edit silently reintroducing the same bug in a slightly different shape.

Net: I'd merge this. It's the right architecture, the bug that got caught during development was a genuine and easy-to-miss one, and it was closed out with a test that actually exercises the failure mode rather than just the fix.
